src/Nonlinear/ReverseAD/Coloring/Coloring.jl

```diff
@@ -0,0 +1,554 @@
+            vec[offset[v_rev]+adjcount[v]] = u_rev
+            adjcount[u] += 1
+            adjcount[v] += 1
+        end
+        resize!(cmap, length(vlist))
+        order, parent =
+            reverse_topological_sort_by_dfs(vec, offset, length(vlist), cmap)
+        for k in 1:length(vlist)
+            revmap[vlist[k]] = 0  # clear for reuse
+        end
+        postorder[idx] = order
+        parents[idx] = parent
+        vertexmap[idx] = vlist
+    end
+    return RecoveryInfo(
+        vertexmap,
+        postorder,
+        parents,
+        color,
+        num_colors,
+        _num_edges(g),
+        local_indices,
+    )
+end
+
+_normalize(i, j) = (j > i) ? (j, i) : (i, j)
+
+function _indirect_recover_structure(rinfo::RecoveryInfo)
+    N = length(rinfo.color)
+    I = zeros(Int, rinfo.nnz + N)
+    J = zeros(Int, rinfo.nnz + N)
+    k = 0
+    for i in 1:N
+        k += 1
+        I[k] = J[k] = i
+    end
+    for t in 1:length(rinfo.postorder)
+        vmap = rinfo.vertexmap[t]
+        order = rinfo.postorder[t]
+        parent = rinfo.parents[t]
+        for z in 1:length(order)
+            v = order[z]
+            p = parent[v]
+            if p == 0
+                continue
+            end
+            k += 1
+            I[k], J[k] = _normalize(vmap[v], vmap[p])
+        end
+    end
+    @assert k == rinfo.nnz + N
+    return I, J
+end
+
+"""
+    hessian_color_preprocess(
+        edgelist,
+        num_total_var,
+        seen_idx = IndexedSet(0),
+    )
+
+`edgelist` contains the nonzeros in the Hessian, *including* nonzeros on the
+diagonal.
+"""
+function hessian_color_preprocess(
+    edgelist,
+    num_total_var,
+    seen_idx = IndexedSet(0),
+)
+    resize!(seen_idx, num_total_var)
+    I, J = Int[], Int[]
+    for (i, j) in edgelist
+        push!(seen_idx, i)
+        push!(seen_idx, j)
+        push!(I, i)
+        push!(J, j)
+    end
+    local_indices = sort!(seen_idx.nzidx[1:seen_idx.nnz])
+    empty!(seen_idx)
+    global_to_local_idx = seen_idx.nzidx # steal for storage
+    for k in 1:length(local_indices)
+        global_to_local_idx[local_indices[k]] = k
+    end
+    # only do the coloring on the local indices
+    for k in 1:length(I)
+        I[k] = global_to_local_idx[I[k]]
+        J[k] = global_to_local_idx[J[k]]
+    end
+    g = UndirectedGraph(I, J, length(local_indices))
+    color, num_colors = acyclic_coloring(g)
+    @assert length(color) == _num_vertices(g)
+    rinfo = recovery_preprocess(g, color, num_colors, local_indices)
+    I, J = _indirect_recover_structure(rinfo)
+    # convert back to global indices
+    for k in 1:length(I)
+        I[k] = local_indices[I[k]]
+        J[k] = local_indices[J[k]]
+    end
+    return I, J, rinfo
+end
+
+"""
+    seed_matrix(rinfo::RecoveryInfo)
+
+Allocate a seed matrix for the Coloring.
+"""
+function seed_matrix(rinfo::RecoveryInfo)
+    return Array{Float64}(undef, length(rinfo.local_indices), rinfo.num_colors)
```

Comment:
```suggestion
    return Matrix{Float64}(undef, length(rinfo.local_indices), rinfo.num_colors)
```
not sure a dedicated function for this is necessary?

Reply:
Maybe not, but it's there. We can refactor Coloring in a different PR.